### PR TITLE
Custom rewrite rule

### PR DESCRIPTION
**Pull request by guillermo136**
Generalizes the grouping functionality to allow for custom rewrite rules. An example of how it works is attached as a screenshot.

image

### CODE REVIEW

File: src/app/input-bar/input-bar.component.ts [hunk 1]
```diff
@@ -37,46 +39,120 @@ export class InputBarComponent implements OnInit {
...
+    // We put each space seperated token in a stack
+    const stackInput = input.trim().split(' ').reverse();
+
+    let currInputArg: string;
+    while (stackInput.length > 0) {
+      currInputArg = stackInput.pop();
+      if (currInputArg === '{') {
+        grouping = this.parseGrouping(stackInput);
+        if (grouping.groupBy === GroupByKeys.Rewrite) {
+          // An empty inputString tells us to not highlight this regex filter
```
Comment (alifradityar) [inline src/app/input-bar/input-bar.component.ts:63]: This sounds like a hack to me. Should we add a new field instead? (e.g. isFromRewrite)

File: src/app/group-sort.pipe.ts [hunk 2]
```diff
@@ -87,14 +106,38 @@ export class GroupSort implements PipeTransform {
...
+  /**
+   * Comparison function for arbitrary objects that compares lexicographically
+   * based on the attributes of the object corresponding to key.
+   * @param { any } first
+   * @param { any } second
+   * @param { string } key The key denoting which attribute should be used for
+   * the comparison.
+   * @return { number }
+   */
+  private lexicographicComp(first: any, second: any, key: string): number {
```
[the remark this answers is not in the document]
Comment (guillermo136) [reply src/app/group-sort.pipe.ts:119]: This is designed to only be used with objects that have specific attribute. For example since we know that all LinkData has an href, we use this with a key of 'href' on LinkData object. If you use this with objects that don't have the provided key then the behavior is undefined.

File: src/app/input-bar/input-bar.component.ts [hunk 3]
```diff
@@ -37,46 +39,120 @@ export class InputBarComponent implements OnInit {
...
+    };
+
+    let currGroupArg: string;
+    while (stackInput.length > 0) {
+      currGroupArg = stackInput.pop();
+      if (currGroupArg === '}') {
+        return grouping;
+      } else if (currGroupArg.startsWith(GroupingModifiers.Sort)) {
+        currGroupArg = currGroupArg.substring(GroupingModifiers.Sort.length);
+        switch (currGroupArg) {
```
[the remark this answers is not in the document]
Comment (guillermo136) [reply src/app/input-bar/input-bar.component.ts:107]: Okay!

File: src/app/group-sort.pipe.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alifradityar) [reply src/app/group-sort.pipe.ts:119]: Right, and it's not preferable to have undefined behavior. I would create a comparator for each type instead of making it generic.

File: src/app/input-bar/input-bar.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alifradityar) [reply src/app/input-bar/input-bar.component.ts:108]: Let me rephrase my previous comment.

I think we should *only* support the following

```
lex-asc
lex-desc
size-asc
size-desc
```

instead of supporting multiple alternatives. 

It's not obvious for user that `ascending` or `asc` is treated as `lexico-asc` using this syntax. It's also not obvious `size` means the same thing as `size-asc`.

[the remark this answers is not in the document]
Comment (guillermo136) [reply src/app/input-bar/input-bar.component.ts:108]: Yeah, that's why I have it set up so that if none of the "officially supported" syntax it will replace it with the official syntax in the input bar to correct the user so that if they input asc, they will see that the parser is treating that as if it were lex-asc. That is what this line does (newInput is the variable bonded to input bar) :

`this.newInput = this.newInput.replace(original, 'sort:lex-desc');`

I though this was the behavior that you described above, let me know if I misunderstood.

File: src/app/group-sort.pipe.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (guillermo136) [reply src/app/group-sort.pipe.ts:119]: okay

File: src/app/input-bar/input-bar.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alifradityar) [reply src/app/input-bar/input-bar.component.ts:108]: Oh you get it correctly. My idea was to do this transformation before "consumeAndParseGrouping()". I wanted to simplify this function a bit since it's doing many things at once. But after looking it a bit more, I think it might complicate things further.

I guess we can keep this logic then
